pkg/storegateway/series_refs_streaming_test.go

```diff
@@ -0,0 +1,353 @@
+	require.NoError(t, it.Err())
+
+	// During chunks streaming phase, the batches should be returned as-is from the new iterator.
+	batches = readAllBatches(it)
+	require.NoError(t, it.Err())
+	require.Equal(t, []seriesChunkRefsSet{firstBatchWithChunkRefs, secondBatchWithChunkRefs}, batches)
+}
+
+func TestChunksStreamingCachingSeriesChunkRefsSetIterator_ExpectingMultipleBatches_AllBatchesFilteredOut(t *testing.T) {
+	psi := newPostingsSetsIterator([]storage.SeriesRef{1, 2, 3, 4, 5, 6}, 3)
+
+	factoryCalls := 0
+	var factoryStrategy seriesIteratorStrategy
+
+	factory := func(strategy seriesIteratorStrategy) iterator[seriesChunkRefsSet] {
+		factoryCalls++
+		factoryStrategy = strategy
+
+		return newSliceSeriesChunkRefsSetIterator(
+			nil,
+			// No batches.
+		)
+	}
+
+	it := newChunksStreamingCachingSeriesChunkRefsSetIterator(defaultStrategy, psi, factory)
+
+	// Inner iterator should be created with chunk refs disabled.
+	require.Equal(t, 1, factoryCalls)
+	require.Equal(t, defaultStrategy.withNoChunkRefs(), factoryStrategy)
+	require.NoError(t, it.Err())
+
+	// Label sending phase.
+	batches := readAllBatches(it)
+	require.NoError(t, it.Err())
+	require.Empty(t, batches)
+
+	// Prepare for chunks streaming phase. Inner iterator should be recreated with chunk refs enabled.
+	it.PrepareForChunksStreamingPhase()
+	require.Equal(t, 2, factoryCalls)
+	require.Equal(t, defaultStrategy.withChunkRefs(), factoryStrategy)
+	require.NoError(t, it.Err())
+
+	// Chunks streaming phase.
+	batches = readAllBatches(it)
+	require.NoError(t, it.Err())
+	require.Empty(t, batches)
+}
+
+func TestChunksStreamingCachingSeriesChunkRefsSetIterator_ExpectingMultipleBatches_IteratorReturnsError(t *testing.T) {
+	psi := newPostingsSetsIterator([]storage.SeriesRef{1, 2, 3, 4, 5, 6}, 3)
+	factoryCalls := 0
+
+	factory := func(_ seriesIteratorStrategy) iterator[seriesChunkRefsSet] {
+		factoryCalls++
+
+		return newSliceSeriesChunkRefsSetIterator(
+			fmt.Errorf("error #%v", factoryCalls),
+			seriesChunkRefsSet{
+				series: []seriesChunkRefs{
+					{lset: labels.FromStrings("series", "1")},
+					{lset: labels.FromStrings("series", "2")},
+					{lset: labels.FromStrings("series", "3")},
+				},
+				releasable: true,
+			},
+		)
+	}
+
+	it := newChunksStreamingCachingSeriesChunkRefsSetIterator(defaultStrategy, psi, factory)
+	require.Equal(t, 1, factoryCalls)
+
+	// During label sending phase, the error from the original iterator should be returned.
+	_ = readAllBatches(it)
+	require.EqualError(t, it.Err(), "error #1")
+
+	// Prepare for chunks streaming phase. Inner iterator should be recreated.
+	it.PrepareForChunksStreamingPhase()
+	require.Equal(t, 2, factoryCalls)
+
+	// During chunks streaming phase, the error from the new iterator should be returned.
+	_ = readAllBatches(it)
+	require.EqualError(t, it.Err(), "error #2")
+}
+
+func readAllBatches(it iterator[seriesChunkRefsSet]) []seriesChunkRefsSet {
```

Comment:
there's already `readAllSeriesChunkRefsSet` which does the same

Reply:
Good catch, removed in 6ca5bc9